### DISCUSSION TITLE
Can Option and Result be used together?

### DISCUSSION THREAD

**Discussion by keiravillekode** [vlang/v#21793] [General]
_Example A_
Suppose we have Lisp/Haskell style lists, and are writing a function to return the tail of the list.

- If the list has one element, we would like to indicate the result is empty, for example by returning `none`.
- If the list has more than one element, we have a non-empty tail, and we need to return a reference.
- If the list has no elements, we would like to report an error.
 

_Example B_
We have a function that constructs binary trees
- An empty tree is a valid result
- A non-empty tree is also a valid result
- If the input is invalid, we want to report an error.

What is the idiomatic way to address such scenarios? Are there examples in the standard library I should look at?

**Comment by JalonSolov** [comment]
No, they cannot.

There used to only be one type, that incorporated both, but too many people complained that it was too "hard" to distinguish.  They actually needed to put an `if` inside the `or` block!

So now, they are 2 separate types which cannot be used together.

The simplest way to get the effects of both is to use a Result type, and send back a specific error for `empty`, etc.  Then check for that specific error in the `or` block of the caller and handle it separately from others.

**Comment by Wajinn** [comment]
V's documentation gives [4 examples](https://github.com/vlang/v/blob/master/doc/docs.md#handling-optionsresults) of usage. Perhaps that should be the reference point to start from, concerning possible issues.
